sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala

```diff
@@ -500,12 +442,98 @@ object GpuDeviceManager extends Logging {
     }
     if (!PinnedMemoryPool.isInitialized && pinnedSize > 0) {
       logInfo(s"Initializing pinned memory pool (${pinnedSize / 1024 / 1024.0} MiB)")
-      PinnedMemoryPool.initialize(pinnedSize, gpuId, setCuioDefaultResource)
+      PinnedMemoryPool.initialize(pinnedSize, gpuId, conf.pinnedPoolCuioDefault)
     }
     // Host memory limits must be set after the pinned memory pool is initialized
     HostAlloc.initialize(nonPinnedLimit)
   }
 
+  // visible for testing
+  def getPinnedPoolAndOffHeapLimits(conf: RapidsConf, sparkConf: SparkConf,
+                                            memCheck: MemoryChecker =
+                                            MemoryCheckerImpl): (Long, Long) = {
+    val perTaskOverhead = conf.perTaskOverhead
+    val totalOverhead = perTaskOverhead * GpuDeviceManager.numCores
+    val confPinnedSize = conf.pinnedPoolSize
+    val confLimit = conf.offHeapLimit
+    // This min limit of 4GB is somewhat arbitrary, but based on some testing which showed
+    // that the previous minimum of 15 MB * num cores was too little for certain benchmark
+    // queries to complete, whereas this limit was sufficient.
+    val minMemoryLimit = 4L * 1024 * 1024 * 1024
+
+    val deviceCount: Int = Cuda.getDeviceCount()
+
+    val executorOverheadKey = "spark.executor.memoryOverhead"
+    val pysparkOverheadKey = "spark.executor.pyspark.memory"
+    val heapSizeKey = "spark.executor.memory"
+
```

Comment:
Do we need to check sparks off heap limit too? `spark.memory.offHeap.size` and `spark.memory.offHeap.enabled`?

I was wrong about this and in Spark 3+ it is not included in the overhead anymore.